R/calculateCoex-method.R

```diff
@@ -843,8 +853,284 @@ contingencyTables <- function(objCOTAN, g1, g2) {
+          logLevel = 3L)
+
+  logThis(paste("Total calculations elapsed time:",
+                difftime(endTime, startTime, units = "secs")),
+          logLevel = 2L)
+
+  logThis(paste("Calculate", kind, "coex (legacy): DONE"), logLevel = 1L)
+
+  return(list("coex" = coex, "ppf" = problematicPairsFraction))
+}
+
+
+# torch based code
+calculateCoex_Torch <- function(objCOTAN, returnPPFract, deviceStr) {
+  logThis(paste0("Calculate genes coex (torch) on device ", deviceStr,
+                 ": START"), logLevel = 1L)
+
+  startTime <- Sys.time()
+
+  gc()
+  cuda_empty_cache()
+
+  # TODO: 16 bits are OK here?
+  if(deviceStr == "cpu"){
+    dtypeForCalc <- torch_float64()
+    halfDtypeForCalc <- torch_float32()
+  } else {
+    dtypeForCalc <- torch_float32()
+    halfDtypeForCalc <- torch_float16()
+  }
+
+  device <- torch_device(deviceStr)
+
+  probOne <- function(nu, lambda, a) {
+    zero     <- torch_tensor( 0.0, device = device, dtype = torch_float64())
+    minusOne <- torch_tensor(-1.0, device = device, dtype = torch_float64())
+
+    # Calculate terms based on conditions
+    term1 <- (a <= zero) *
+      torch_exp(torch_ger(nu, lambda) * (torch_minimum(a, zero) + minusOne))
+
+    term2 <- (a >  zero) *
+      torch_pow(one + torch_ger(nu, lambda) * torch_maximum(a, zero),
+                minusOne / a)
+
+    invisible(term1$add_(term2)$add_(minusOne)$neg_())
+
+    return(term1)
+  }
+
+  logThis("Retrieving expected genes contingency table", logLevel = 3L)
+
+  one <- torch_tensor(1.0, device = device, dtype = dtypeForCalc)
+  m   <- torch_tensor(getNumCells(objCOTAN),
+                      device = device, dtype = dtypeForCalc)
+
+  expectedYY <-torch_tensor(
+    probOne(torch_tensor(getNu(objCOTAN),
+                         dtype = torch_float64(), device = device),
+            torch_tensor(getLambda(objCOTAN),
+                         dtype = torch_float64(), device = device),
+            torch_tensor(getDispersion(objCOTAN),
+                         dtype = torch_float64(), device = device)),
+    device = device, dtype = dtypeForCalc)
+  cuda_empty_cache()
+
+  expectedY <- torch_sum(expectedYY, 1L, dtype = dtypeForCalc)
+
+  expectedYY <- torch_matmul(torch_t(expectedYY), expectedYY)
+
+  expectedTime <- Sys.time()
+  logThis(paste("Expected genes contingency table elapsed time:",
+                difftime(expectedTime, startTime, units = "secs")),
+          logLevel = 3L)
+
+  logThis("Calculating genes coex normalization factor", logLevel = 3L)
+
+  coex <- torch_maximum(expectedYY, one)$reciprocal_()
+
+  thresholdForPP <- 0.5
+  problematicPairs <- NULL
+  if (isTRUE(returnPPFract)) {
+    problematicPairs <-
+      torch_tensor(expectedYY < thresholdForPP,
+                   dtype = torch_bool(), device = device)
+  }
+
+  # expectedYN
+  tmp <- expectedY$view(c(-1L, 1L)) - expectedYY
+  invisible(coex$add_(torch_maximum(tmp, one)$reciprocal_()))
```

Comment:
Add a TODO: use in-place torch_maximum_() as soon as it becomes available

Reply:
Done